compiler/rustc_trait_selection/src/traits/query/type_op/mod.rs

```diff
@@ -122,39 +129,9 @@ where
 
     fn fully_perform(self, infcx: &InferCtxt<'_, 'tcx>) -> Fallible<TypeOpOutput<'tcx, Self>> {
         let mut region_constraints = QueryRegionConstraints::default();
-        let (output, canonicalized_query, mut obligations, _) =
+        let (output, canonicalized_query) =
             Q::fully_perform_into(self, infcx, &mut region_constraints)?;
 
-        // Typically, instantiating NLL query results does not
-        // create obligations. However, in some cases there
-        // are unresolved type variables, and unify them *can*
-        // create obligations. In that case, we have to go
-        // fulfill them. We do this via a (recursive) query.
-        while !obligations.is_empty() {
-            trace!("{:#?}", obligations);
-            let mut progress = false;
-            for obligation in std::mem::take(&mut obligations) {
```

Comment:
I wonder if this hurts perf. We create a new Vec when we take, so we have to reallocate when new obligations are pushed.

Reply:
trying this out in https://github.com/rust-lang/rust/pull/93007